sdks/python/apache_beam/runners/portability/fn_api_runner/fn_runner.py

```diff
@@ -845,10 +845,12 @@ def process_bundle(self,
           (result_future.is_done() and result_future.get().error)):
         if isinstance(output, beam_fn_api_pb2.Elements.Timers) and not dry_run:
           with BundleManager._lock:
-            self.bundle_context_manager.get_buffer(
+            timer_buffer = self.bundle_context_manager.get_buffer(
                 expected_output_timers[(
                     output.transform_id, output.timer_family_id)],
-                output.transform_id).append(output.timers)
+                output.transform_id)
+            timer_buffer.cleared = False
```

Comment:
I don't think we should set `timer_buffer.cleared` to `False` here.  There is an assumption around `FnApiRunner.ListBuffer` that per key the timer is only fired once. It seems like the assumption is broken by setting a new timer when firing the timer. If that's the case, we should consider updating the logic around timer buffer.

Reply:
Where can I find this assumption? Note that the tests are passing.

I think it makes sense to reset the flag; when the timer is set after firing we should treat it independently from any prior settings of the timer. We may want to leave a comment here why this is necessary. Alternatively, the resetting may be done directly after the timer is fired.

Reply:
@pabloem who is the original author of `ListBuffer`. I prefer the second way Max mentioned above. What do you think? Other than this part, timer changes look good to me.

Reply:
Did you run into these errors when setting a timer from the timer call, Max? I think it would be good to explicitly reset the buffer, rather than manipulate its flag (e.g. write a timer_buffer.reset function). Or at least check `if timer_buffer.cleared: timer_buffer.cleared = False`, to confirm that the rest of the internal context in `ListBuffer` is cleared.

Reply:
>Did you run into these errors when setting a timer from the timer call, Max?

That's the point of this PR :) It is simply not possible to set a new timer from a firing timer without resetting the timer buffer.


>  Or at least check `if timer_buffer.cleared: timer_buffer.cleared = False`

That is semantically identical to the current code. It is redundant to have the if check because when `timer_buffer.cleared` is `False`, then setting it to False again won't have any effect.

I can add a `reset()` method which additionally checks whether the buffer had been cleared before. I think this is what you meant by the check.

Reply:
Updated. PTAL @pabloem

Reply:
This made things more complicated because I also have to allow resetting when the buffer has never been used before, i.e. when the timer set for the first time.

Reply:
PTAL @pabloem